### DISCUSSION TITLE
Nextjs behind url rewriting Lambda@Edge

### DISCUSSION THREAD

**Discussion by nihakue** [vercel/next.js#16572] [Help]
I would like to deploy a nextjs behind a Lambda@Edge (CloudFront edge worker) function that does some authorization, reads a cookie, and then routes requests to the origin based on the results. This way, the user can request urls like this:

`/posts/:postId`

that are forwarded to my nextjs origin server as: 

`/:language/:permission/posts/:postId`

This has the advantage of only ever exposing 'clean' urls to users, while still getting good cache performance since the CDN caches the expanded url, rather than the clean one.

With this setup, when I load a SSRd page everything seems to work OK, but I can't get client-side routing to work since the 'clean' urls are not in my route manifests (I believe) and my hrefs don't match my 'as' props.

An example link:

```jsx
<Link href="/[language]/[permission]/posts/[postId]" as="/posts/1234"><a>Click</a></Link>
```

I am also having various problems with SSG/ISR, but I suspect that they are related to the issues with SSR, so I'll focus on those for now.

This setup isn't all that different to putting nextjs behind a reverse proxy, so I'm hoping that I'm missing something obvious and there is a way to support this without resorting to modifying all of my route manifests as a post-build step (I'm not even sure that would work).

Does anyone have any advice?

**Comment by conanbatt** [comment]
Where you able to solve this issue?

Since next 9.5.3 the "as" prop is no longer needed, but I don't know if it would resolve your problem.

**Reply by nihakue** [reply]
I'm still struggling with this, I haven't tried 9.5.3 to see if the new Link/Router solve my issue. I'll do that now and update.

**Reply by nihakue** [reply]
My main issue right now (which isn't effected by the 9.5.3 changes afaict) is that nextJS treats `/posts/1` as an 'external URL' and does a full page navigation. If there were some way to provide a mapping from `/posts/1` to `/[language]/[permission]/posts/[postId]` (and some other means of filling in the dynamic parts) so that next was aware that this was an internal route and could use client-side navigation I think that would solve my problems.

**Reply by nihakue** [reply]
I think I've solved (or at least successfully worked around) this issue. I'm using optional catch-all routes instead of dynamic routes or query parameter,

so instead of `/[language]/[permission]/posts/[postId]` have `/posts/[[...props]].tsx` which parses {props: []} into the components parts. For example, when /posts/en-us/admin/12 becomes `{props: ['en-us', 'admin', '12']}`

This works for static and SSR because the same template (html) page is used regardless of the props, and the url matches the 'internal URL' rules on the client side so client-side routing, pre-fetching, etc work. 

So now all I do is proxy /posts/12 requests, read the permission/locale values from the cookie (or whatever), and forward to the full route. The client is happy to have window.location.href = `/posts/12` because that matches `/posts/[[...props]].tsx`

Honestly it feels like a bit of a hack, but it works for me. I feel like statically rendered pages should just support query params (since how are they any different that optional catch-all routes?)
